source/slang/slang-check-decl.cpp

```diff
@@ -7096,13 +7096,13 @@ namespace Slang
             // Two such constraints are equivalent if their `sub`
             // and `sup` types are pairwise equivalent.
             //
-            auto leftSub = leftConstraint->sub;
-            auto rightSub = getSub(m_astBuilder, rightConstraint);
+            auto leftSub = leftConstraint->sub.type;
+            auto rightSub = substInnerRightToLeft.substitute(m_astBuilder, rightConstraint.getDecl()->sub.type);
```

Comment:
I thought `rightConstraint` already contains everything needed to specialize `sub` and `sup`, why isn't that working?

Reply:
The issue is that when referring directly to `GenericTypeConstraintDecl` or `GenericTypeParam` we don't construct the substitutions tree at all. There's special-case logic that always uses the direct-decl-ref. As a result, there are never any substitutions on the constraint decl-ref.